scipy/integrate/_quadrature.py

```diff
@@ -662,6 +664,216 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+            ],
+            axis=0,
+        )
+    else:
+        sub_integrals = np.concatenate(
+            [
+                sub_integrals_fwd[0:1],
+                (sub_integrals_fwd[1:] + sub_integrals_bwd[:-1]) / 2,
+                sub_integrals_bwd[-1:],
+            ],
+            axis=0,
+        )
+
+    res = np.cumsum(sub_integrals, axis=0).swapaxes(0, axis)
+
+    return res
+
+
+def cumulative_simpson(y, x=None, dx=1.0, axis=-1, initial=None):
+    """
+    Cumulatively integrate y(x) using Simpson's 1/3 rule. The integral of the samples
+    at every point is calculated by assuming a quadratic relationship between each point and two
+    other adjacent points.
+
+    Parameters
+    ----------
+    y : array_like
+        Values to integrate.
+    x : array_like, optional
+        The coordinate to integrate along. Must be monotonically increasing.
+        If None (default), use spacing `dx` between consecutive elements in `y`.
+    dx : scalar | array_like, optional
+        Spacing between elements of `y`. Only used if `x` is None. Can either be
+        a float, or an array with the same shape as `y`, but size 1 along `axis`. 
+    axis : int, optional
+        Specifies the axis to integrate along. Default is -1 (last axis).
+    initial : scalar | array_like, optional
+        If given, insert this value at the beginning of the returned result,
+        and add it to the rest of the result. Typically this value should be 0. 
+        Default is None, which means no value at ``x[0]`` is returned and `res` 
+        has one element less than `y` along the axis of integration.
+        Can either be a float, or an array with the same shape as `y`, 
+        but size 1 along `axis`. 
+
+    Returns
+    -------
+    res : ndarray
+        The result of cumulative integration of `y` along `axis`.
+        If `initial` is None, the shape is such that the axis of integration
+        has one less value than `y`. If `initial` is given, the shape is equal
+        to that of `y`.
+
+    See Also
+    --------
+    numpy.cumsum, numpy.cumprod
+    cumulative_trapezoid : cumulative integration using the composite trapezoidal rule
+    quad : adaptive quadrature using QUADPACK
+    romberg : adaptive Romberg quadrature
+    quadrature : adaptive Gaussian quadrature
+    fixed_quad : fixed-order Gaussian quadrature
+    dblquad : double integrals
+    tplquad : triple integrals
+    romb : integrators for sampled data
+    ode : ODE integrators
+    odeint : ODE integrators
+
+    Notes
+    -----
+
+    .. versionadded:: 1.11.0
+
+    Examples
+    --------
+    >>> from scipy import integrate
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+
+    >>> x = np.linspace(-2, 2, num=20)
+    >>> y = x
+    >>> y_int = integrate.cumulative_simpson(y, x, initial=0)
+    >>> plt.plot(x, y_int, 'ro', x, y[0] + 0.5 * x**2, 'b-')
+    >>> plt.show()
+
+    """
+
+    y = np.asarray(y).astype(np.float64)
```

Comment:
Why does it need to be cast to float64?

Reply:
The dtypes were cast to ensure all inputs (y, x, dx, initial) are of the same type. This way there is no chance of an inaccurate result where a float array is output as an int32. 

For eg in [line 872](https://github.com/scipy/scipy/pull/18151/files#diff-2ff88b8913ad74b9ea3fe1fe02d6d49164855189168b767076af2e94206c3a2cR872), if initial is of dtype int32, and res is dtype float64, then the output of the function will still be int32.

Casting the dtypes from the outset would ensure the function gives consistent results.

What do you think?

Reply:
Thanks for the recommendations @MatteoRaso . I have now removed the casting.

I also updated the formulae so that the term coefficients are [floats](https://github.com/scipy/scipy/pull/18151/commits/cbc46b5f7944db8b87c18f919a1d153edcc0f4f0#). This ensures that even if the inputs are all of type `int`, the output will be in `float` (consistent with cumulative_trapezoid).

Earlier i was having an issue with `initial[slice1] += res`. This was giving an error if initial was `np.int32` and res was `np.float64`. This had led me to explicitly casting, but `initial[slice1] = initial[slice1] + res` does the type casting correctly.